ff/serving/provider/snowflake.go

```diff
@@ -0,0 +1,626 @@
+	var idType string
+	if id.Type == Feature {
+		idType = "feature"
+	} else {
+		idType = "label"
+	}
+	return fmt.Sprintf("featureform_resource_%s_%s_%s", idType, id.Name, id.Variant)
+}
+
+func (store *snowflakeOfflineStore) getMaterializationTableName(ftID MaterializationID) string {
+	return fmt.Sprintf("featureform_materialization_%s", ftID)
+}
+
+func (store *snowflakeOfflineStore) getTrainingSetName(id ResourceID) string {
+	return fmt.Sprintf("featureform_trainingset_%s_%s", id.Name, id.Variant)
+}
+
+func (store *snowflakeOfflineStore) tableExists(id ResourceID) (bool, error) {
+	n := -1
+	var tableName string
+	if id.check(Feature, Label) == nil {
+		tableName = store.getResourceTableName(id)
+	} else if id.check(TrainingSet) == nil {
+		tableName = store.getTrainingSetName(id)
+	}
+	err := store.db.QueryRow(`SELECT COUNT(*) FROM information_schema.tables WHERE table_name = ?`, tableName).Scan(&n)
+	if n == 0 {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (store *snowflakeOfflineStore) AsOfflineStore() (OfflineStore, error) {
+	return store, nil
+}
+
+// CreateResourceTable creates a new Resource table.
+// Returns a table if it does not already exist and stores the table ID in the resource index table.
+// Returns an error if the table already exists or if table is the wrong type.
+func (store *snowflakeOfflineStore) CreateResourceTable(id ResourceID, schema SerializedTableSchema) (OfflineTable, error) {
+	psSchema := SnowflakeSchema{}
+	if err := psSchema.Deserialize(schema); err != nil {
+		return nil, err
+	}
+	if err := id.check(Feature, Label); err != nil {
+		return nil, err
+	}
+
+	if exists, err := store.tableExists(id); err != nil {
+		return nil, err
+	} else if exists {
+		return nil, &TableAlreadyExists{id.Name, id.Variant}
+	}
+	tableName := store.getResourceTableName(id)
+	table, err := newSnowflakeOfflineTable(store.db, tableName, psSchema.ValueType)
+	if err != nil {
+		return nil, err
+	}
+	return table, nil
+}
+
+func (store *snowflakeOfflineStore) GetResourceTable(id ResourceID) (OfflineTable, error) {
+	return store.getSnowflakeResourceTable(id)
+}
+
+type snowflakeMaterialization struct {
+	id        MaterializationID
+	db        *sql.DB
+	tableName string
+}
+
+func (mat *snowflakeMaterialization) ID() MaterializationID {
+	return mat.id
+}
+
+func (mat *snowflakeMaterialization) NumRows() (int64, error) {
+	n := int64(0)
+	query := fmt.Sprintf("SELECT COUNT(*) FROM %s", sanitize(mat.tableName))
+	rows := mat.db.QueryRow(query)
+	err := rows.Scan(&n)
+	if err != nil {
+		return 0, err
+	}
+	return n, nil
+}
+
+func (mat *snowflakeMaterialization) IterateSegment(start, end int64) (FeatureIterator, error) {
+	query := fmt.Sprintf(""+
+		"SELECT entity, value, ts FROM "+
+		"( SELECT * FROM "+
+		"( SELECT *, row_number() over(ORDER BY (SELECT NULL)) as row_number FROM %s )t1 WHERE row_number>? AND row_number<=?)t2", sanitize(mat.tableName))
```

Comment:
this doesn't guarantee an order, so is it guaranteed that there will be no overlapping between segments given the order by isn't deterministic? If this is verified with testing ignore this comment